### PR TITLE
Get one document

### PR DESCRIPTION
**Pull request by jennyesoo**
104250 - Write code : update one document function

### CODE REVIEW

File: HP.Pulsar.Search.Keyword/DataReader/ChangeRequestReader.cs [hunk 1]
```diff
@@ -13,9 +13,54 @@ public ChangeRequestReader(KeywordSearchInfo info)
         _info = info;
     }
 
-    public Task<CommonDataModel> GetDataAsync(int changeRequestId)
+    public async Task<CommonDataModel> GetDataAsync(int changeRequestId)
     {
-        throw new NotImplementedException();
+        IEnumerable<CommonDataModel> changeRequest = await GetChangeRequestAsync(changeRequestId);
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataReader/ChangeRequestReader.cs:18]: There is a bug here.
The collection may contain no element when change request id doesn't exist in database.
When the collection is empty, you should not continue to process the collection and then you should return null. 
Also, you need to make sure that callers can handle null object well.

File: HP.Pulsar.Search.Keyword/DataReader/ComponentRootReader.cs [hunk 2]
```diff
@@ -13,9 +13,58 @@ public ComponentRootReader(KeywordSearchInfo info)
         _info = info;
     }
 
-    public Task<CommonDataModel> GetDataAsync(int componentRootId)
+    public async Task<CommonDataModel> GetDataAsync(int componentRootId) 
     {
-        throw new NotImplementedException();
+        IEnumerable<CommonDataModel> componentRoot = await GetComponentRootAsync(componentRootId);
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataReader/ComponentRootReader.cs:18]: Same problem here.
I don't repeat the same problem again in the following code.
You have to update all of them.

File: HP.Pulsar.Search.Keyword/DataReader/ChangeRequestReader.cs [hunk 3]
```diff
@@ -13,9 +13,54 @@ public ChangeRequestReader(KeywordSearchInfo info)
...
+        {
+            HandlePropertyValuesAsync(changeRequest),
+            GetApproverAsync(changeRequest)
+        };
+
+        await Task.WhenAll(tasks);
+        return changeRequest.First();
+    }
+
+    private async Task<IEnumerable<CommonDataModel>> GetChangeRequestAsync(int changeRequestId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataReader/ChangeRequestReader.cs:30]: I don't think this is a good function.
When an id was entered, we expected that only one element will be returned. So the return type should not be "collection" type. It should be justCommonDataModel.

Fix the problem for all of the code in other target.

**Review by wooly905 — CHANGES_REQUESTED** [review]
Please update the code and then sent a new PR.

**Comment by jennyesoo** [comment]
Update:
1. 新增UpdateClient，根據使用者輸入的SearchType去更新Type項目，並呼叫Type的Orchestrator。
2. 每個Orchestrator, Reader, Transformer已多新增給<CommonDataModel>使用的function。
3. 已新增當CommonDataModel為空項目時，就不進行之後動作。

File: HP.Pulsar.Search.Keyword/DataReader/ProductReader.cs [hunk 4]
```diff
@@ -617,12 +983,52 @@ private async Task GetAvDetailAsync(IEnumerable<CommonDataModel> products)
         }
     }
 
+    private async Task GetFactoryNameAsync(CommonDataModel product)
+    {
+        if (int.TryParse(product.GetValue("ProductId"), out int productId))
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataReader/ProductReader.cs:988]: same here, code style.

File: HP.Pulsar.Search.Keyword/DataTransformation/ChangeRequestDataTransformer.cs [hunk 5]
```diff
@@ -29,4 +29,23 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> chang
 
         return changeRequests;
     }
+
+    public CommonDataModel Transform(CommonDataModel changeRequest)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/ChangeRequestDataTransformer.cs:33]: Input can be null. You need to handle this case.

All public functions must verify all arguments to make sure that they are valid to be processed in this function.
This is coding 101. You need to keep in mind for always.

File: HP.Pulsar.Search.Keyword/DataTransformation/ComponentRootTransformer.cs [hunk 6]
```diff
@@ -28,4 +28,23 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> compo
 
         return componentRoots;
     }
+
+    public CommonDataModel Transform(CommonDataModel componentRoot)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/ComponentRootTransformer.cs:32]: Same problem here, verify input.

File: HP.Pulsar.Search.Keyword/DataTransformation/ComponentVersionDataTransformer.cs [hunk 7]
```diff
@@ -28,4 +28,24 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> compo
 
         return componentVersions;
     }
+
+    public CommonDataModel Transform(CommonDataModel rootversion)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/ComponentVersionDataTransformer.cs:32]: Same here. verify input.

argument name -> version 
Don't say rootversion, I can't tell it's root or version.

File: HP.Pulsar.Search.Keyword/DataTransformation/FeatureDataTransformer.cs [hunk 8]
```diff
@@ -28,4 +28,24 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> featu
 
         return features;
     }
+
+    public CommonDataModel Transform(CommonDataModel feature)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/FeatureDataTransformer.cs:32]: Same problem here.

File: HP.Pulsar.Search.Keyword/DataTransformation/HpAMOPartNumberDataTransformer.cs [hunk 9]
```diff
@@ -28,4 +28,23 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> hpPar
 
         return hpPartNumber;
     }
+
+    public CommonDataModel Transform(CommonDataModel partNumber)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/HpAMOPartNumberDataTransformer.cs:32]: Same problem here

File: HP.Pulsar.Search.Keyword/DataTransformation/ProductDataTransformer.cs [hunk 10]
```diff
@@ -28,4 +28,23 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> produ
 
         return products;
     }
+
+    public CommonDataModel Transform(CommonDataModel product)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/ProductDataTransformer.cs:32]: Same problem here

File: HP.Pulsar.Search.Keyword/DataTransformation/ProductDropDataTransformer.cs [hunk 11]
```diff
@@ -23,4 +23,19 @@ public IEnumerable<CommonDataModel> Transform(IEnumerable<CommonDataModel> produ
 
         return productDrop;
     }
+
+    public CommonDataModel Transform(CommonDataModel productDrop)
+    {
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/DataTransformation/ProductDropDataTransformer.cs:28]: Same problem here.

File: HP.Pulsar.Search.Keyword/Orchestrator/ChangeRequestOrchestrator.cs [hunk 12]
```diff
@@ -33,4 +33,22 @@ public async Task InitializeAsync()
 
         await mClient.SendElementsCreationAsync(changeRequests);
     }
+
+    public async Task InitializeAsync(int changeRequestId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/ChangeRequestOrchestrator.cs:37]: I don't think we need this. Init is to make everything from scratch, which means Meilisearch is empty. Init will read many things from PRS and write the things to Meilisearch.
We don't init one single thing. It's not called init. It's "update".
We will implement UpdateClient later. So remove the function now.

File: HP.Pulsar.Search.Keyword/Orchestrator/ComponentRootOrchestrator.cs [hunk 13]
```diff
@@ -33,4 +33,23 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(roots);
     }
+
+    public async Task InitializeAsync(int rootId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/ComponentRootOrchestrator.cs:37]: Same too

File: HP.Pulsar.Search.Keyword/Orchestrator/ComponentVersionOrchestrator.cs [hunk 14]
```diff
@@ -33,4 +33,22 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(versions);
     }
+
+    public async Task InitializeAsync(int componentVersionId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/ComponentVersionOrchestrator.cs:37]: Same too

File: HP.Pulsar.Search.Keyword/Orchestrator/FeatureOrchestrator.cs [hunk 15]
```diff
@@ -33,4 +33,22 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(features);
     }
+
+    public async Task InitializeAsync(int featureId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/FeatureOrchestrator.cs:37]: Same too

File: HP.Pulsar.Search.Keyword/Orchestrator/HpAMOPartNumberOrchestrator.cs [hunk 16]
```diff
@@ -33,5 +33,23 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(hpAMOPartNumber);
     }
+
+    public async Task InitializeAsync(int hpAMOPartNumberId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/HpAMOPartNumberOrchestrator.cs:37]: Same too

File: HP.Pulsar.Search.Keyword/Orchestrator/IInitializationOrchestrator.cs [hunk 17]
```diff
@@ -6,4 +6,6 @@ internal interface IInitializationOrchestrator
 {
     KeywordSearchInfo KeywordSearchInfo { get; }
     public Task InitializeAsync();
+    public Task InitializeAsync(int id);
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/IInitializationOrchestrator.cs:9]: Remove

File: HP.Pulsar.Search.Keyword/Orchestrator/ProductDropOrchestrator.cs [hunk 18]
```diff
@@ -33,4 +33,24 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(productDrops);
     }
+
+    public async Task InitializeAsync(int productDropId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/ProductDropOrchestrator.cs:37]: Remove

File: HP.Pulsar.Search.Keyword/Orchestrator/ProductOrchestrator.cs [hunk 19]
```diff
@@ -33,4 +33,22 @@ public async Task InitializeAsync()
 
         await writer.SendElementsCreationAsync(products);
     }
+
+    public async Task InitializeAsync(int productId)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/Orchestrator/ProductOrchestrator.cs:37]: Remove

File: HP.Pulsar.Search.Keyword/SearchEngine/MeiliSearchClient.cs [hunk 20]
```diff
@@ -71,6 +71,27 @@ public async Task SendElementsCreationAsync(IEnumerable<CommonDataModel> element
         await index.AddDocumentsAsync(pairs);
     }
 
+    /// <summary>
+    /// This method is only for sending signal to meilisearch to update a new element.
+    /// When this method completes, it doesn't mean the new element has been updated in meilisearch.
+    /// </summary>
+    /// <param name="elements"></param>
+    /// <exception cref="ArgumentException"></exception>
+    public async Task SendElementsUpdateAsync(CommonDataModel element)
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/SearchEngine/MeiliSearchClient.cs:80]: Method name -> SendElementUpdationAsync

File: HP.Pulsar.Search.Keyword/UpdateEngine/UpdateClient.cs [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    private readonly KeywordSearchInfo _keywordSearchInfo;
+    private IInitializationOrchestrator? _item;
+
+    public UpdateClient(KeywordSearchInfo info)
+    {
+        _keywordSearchInfo = info;
+    }
+
+    public async Task UpdateAsync(SearchType searchType, int id)
+    {
```
Comment (wooly905) [inline HP.Pulsar.Search.Keyword/UpdateEngine/UpdateClient.cs:25]: Remove this file. We are not here yet.

File: KeywordSearchInitializationConsole/Program.cs [hunk 22]
```diff
@@ -21,19 +20,15 @@ private static async Task Main(string[] args)
...
 
-        //DateTime end = DateTime.Now;
-        //Console.WriteLine("total seconds = " + (end - start).TotalSeconds);
-
-        // update
-        //Update update = new();
-        //update.update(HP.Pulsar.KeywordSearch.CommonDataStructures.SearchType.Product, 2018);
-
+        //update
+        UpdateClient updateClient = new UpdateClient(info);
```
Comment (wooly905) [inline KeywordSearchInitializationConsole/Program.cs:31]: Keep them commented. We don't implement it now.

File: KeywordSearchInitializationConsole/Program.cs [hunk 23]
```diff
@@ -4,15 +4,14 @@
 using HP.Pulsar.Search.Keyword.Infrastructure;
 using HP.Pulsar.Search.Keyword.Orchestrator;
 using HP.Pulsar.Search.Keyword.SearchEngine;
+using HP.Pulsar.Search.Keyword.UpdateEngine;
```
Comment (wooly905) [inline KeywordSearchInitializationConsole/Program.cs:7]: Remove

**Review by wooly905 — APPROVED** [review]
Resolve all comments and then merge to dev.
If any, ping me.

**Review by wooly905 — APPROVED** [review]
Resolve all comments and then merge to dev.
If any, ping me.

**Comment by wooly905** [comment]
@jennyesoo 
Can you please resolve the conflict and merge again?

**Comment by jennyesoo** [comment]
> @jennyesoo Can you please resolve the conflict and merge again?

Sure, I am currently working on it.

**Pull request merged**